lib/miq_automation_engine/service_models/miq_ae_service_container_deployment.rb

```diff
@@ -0,0 +1,172 @@
+      object_send(:generate_ansible_inventory_for_subscription)
+    end
+
+    def add_automation_task(task)
+      ar_method do
+        wrap_results(@object.automation_task = AutomationTask.find_by_id(task.id))
+        @object.save!
+      end
+    end
+
+    def customize(options = {})
+      ar_method do
+        if @object.customizations.empty?
+          @object.customizations = {:agent => {}}
+        end
+        options.each do |key, val|
+          wrap_results(@object.customizations[:agent][key] = val)
+        end
+        @object.save!
+      end
+    end
+
+    def perform_scp(ip, username, local_path, remote_path)
+      Net::SCP.upload!(ip,
+                       username,
+                       local_path,
+                       remote_path,
+                       :ssh => {:key_data => ssh_auth.auth_key})
+    end
+
+    def perform_agent_commands(ip, username, commands = [])
+      pp commands
+      ssh = LinuxAdmin::SSH.new(ip, username, ssh_auth.auth_key)
+      result = ssh.perform_commands(commands)
+      result
+    end
+
+    def check_connection(ip, username, ips)
+      agent = LinuxAdmin::SSHAgent.new(ssh_auth.auth_key, "/tmp/ssh_manageiq/ssh_manageiq_#{id}")
+      success, unreachable_ips = agent.with_service do |socket|
+        ssh = LinuxAdmin::SSH.new(ip, username)
+        unreachable_ips = []
+        conneted = true
+        ips.each do |sub_ip|
+          res = ssh.perform_commands(["sudo -E ssh -o 'StrictHostKeyChecking no' #{username}@#{sub_ip} echo $?"], socket)[:stdout]
+          unless res.include?("0\r\n")
+            conneted = false
+            unreachable_ips << sub_ip
+          end
+        end
+        [conneted, unreachable_ips]
+      end
+      unless success
+        raise StandardError, "couldn't connect to : #{unreachable_ips.join(',')}"
+      end
+      success
+    end
+
+    def playbook_running?
+      pid = ar_method do
+        @object.customizations[:agent][:agent_pid] if @object.customizations[:agent]
+      end
+      !pid.nil?
+    end
+
+    def run_playbook_command(ip, username, cmd)
+      result = {:finished => false}
+      playbook_pid = ar_method do
+        @object.customizations[:agent][:playbook_pid] if @object.customizations[:agent]
+      end
+      ssh = LinuxAdmin::SSH.new(ip, username, ssh_auth.auth_key)
+      if playbook_pid
+        process_running = ssh.perform_commands(["sudo kill -0 #{playbook_pid}"])[:stdout]
+        unless process_running.empty?
+          result[:finished] = true
+          result[:stdout] = ssh.perform_commands(["sudo cat /tmp/ansible.log"])[:stdout]
+          stop_agent(ssh)
+        end
+      else
+        pid, socket = create_agent(ssh, username)
+        ssh.perform_commands(["sudo rm -f /tmp/*ansible.log*"])
+        ssh.perform_commands(["SSH_AUTH_SOCK=#{socket} nohup sudo -b -E #{cmd}"])
+        playbook_pid = ssh.perform_commands(["sudo pgrep -f 'sudo -b -E'"])[:stdout].split("\r").first
```

Comment:
I don't like this grep.
1. There could be wholly unrelated `sudo -b -E` on the machine.
2. Doesn't `sudo` eventually exec to given command and disappear from ps?  Perhaps not, but I'd consider that sudo's implementation detail.

Since executing ansible is so fragile, I'm :+1: on committing *something* and iterating in later PRs, which would leave a history trail.